File: mandelbrot.hs
-- Configuration
width = 180
height = 40
maxIterations = 250
xBounds = Bounds (-2.0, 2.0)
yBounds = Bounds (-1.0, 1.0)
center = Complex (0.0, 0.0)

xScale = (bMax xBounds + bMin xBounds) / width
yScale = (bMax yBounds + bMin yBounds) / height

mandelbrot :: Complex -> Complex -> Integer -> Bool
mandelbrot c z i
    | i >= maxIterations = True
    | re z3 * re z3 + im z3 * im z3 > 4.0 = False
    | otherwise = mandelbrot c z3 (i + 1)
  where
    z2 =
        Complex
            ( re z * re z - im z - im z
            , 2.0 * re z * im z
            )
    z3 =
        Complex
            ( re z2 + re c
            , im z2 + im c
            )

-- Complex Type & functions
data Complex = Complex (Double, Double)

re :: Complex -> Double
re (Complex (re, _)) = re

im :: Complex -> Double
im (Complex (_, im)) = im

-- Bounds Type & functions
data Bounds = Bounds (Double, Double)

bMin :: Bounds -> Double
bMin (Bounds (min, _)) = min

bMax :: Bounds -> Double
bMax (Bounds (_, max)) = max

File: mandelbrot.js
const cfg = {
  xBounds: { min: -2.0, max: 2.0 },
  yBounds: { min: -1.0, max: 1.0 },
  center: { re: -0.5, im: 0 },
  maxIterations: 1000,
  height: 40,
  width: 180,
}

const xScale = (cfg.xBounds.max - cfg.xBounds.min) / cfg.width
const yScale = (cfg.yBounds.max - cfg.yBounds.min) / cfg.height

function mandelbrot(c) {
  let z = { re: 0, im: 0 }
  let iterCount = 0

  while (z.re + z.im <= 4 && iterCount < cfg.maxIterations) {
    z = {
      re: z.re * z.re - z.im * z.im,
      im: 2 * z.re * z.im,
    }
    z = {
      re: z.re + c.re,
      im: z.im + c.im,
    }
    iterCount++
  }

  return (iterCount === cfg.maxIterations)
}

function plot() {
  let grid = new Array(cfg.height);

  for (let row = 0; row < cfg.height; row++) {
    grid[row] = new Array(cfg.width)

    for (let col = 0; col < cfg.width; col++) {
      c = {
        re: col * xScale + cfg.xBounds.min + cfg.center.re,
        im: row * yScale + cfg.yBounds.min + cfg.center.im,
      }
      grid[row][col] = mandelbrot(c) ? "*" : " "
    }
  }

  for (let row = 0; row < cfg.height; row++) {
    grid[row] = grid[row].join('')
  }

  return grid.join("\n")
}

console.log(plot())
